src/lib/mina.ts

```diff
@@ -163,69 +205,151 @@ export function LocalBlockchain({
     testAccounts.push({ privateKey: k, publicKey: pk });
   }
 
-  function getAccount(pk: PublicKey): Account {
-    const r = ledger.getAccount(pk);
-    if (r == undefined) {
-      throw new Error(
-        `getAccount: Could not find account for ${JSON.stringify(pk.toJSON())}`
+  return {
+    accountCreationFee: () => UInt32.fromString(accountCreationFee_),
+    currentSlot() {
+      return UInt32.fromNumber(
+        Math.ceil((new Date().valueOf() - startTime) / msPerSlot)
       );
-    } else {
-      const a = {
-        balance: new UInt64(r.balance.value),
-        nonce: new UInt32(r.nonce.value),
-        zkapp: r.zkapp,
-      };
-      return a;
-    }
-  }
-
-  function transaction(sender: PrivateKey, f: () => unknown) {
-    let txn = createTransaction(sender, f);
-    return {
-      ...txn,
-      send() {
-        txn.sign();
-        ledger.applyPartiesTransaction(
-          toParties(txn.transaction),
-          accountCreationFee_
+    },
+    getAccount(publicKey: PublicKey): Account {
+      let ledgerAccount = ledger.getAccount(publicKey);
+      if (ledgerAccount == undefined) {
+        throw new Error(
+          `getAccount: Could not find account for public key ${publicKey.toBase58()}`
         );
-        return { wait: async () => {} };
-      },
-    };
-  }
-
-  function applyJsonTransaction(json: string) {
-    return ledger.applyJsonTransaction(json, accountCreationFee_);
-  }
+      } else {
+        return {
+          publicKey: publicKey,
+          balance: new UInt64(ledgerAccount.balance.value),
+          nonce: new UInt32(ledgerAccount.nonce.value),
+          zkapp: ledgerAccount.zkapp,
+        };
+      }
+    },
+    sendTransaction(txn: Transaction) {
+      txn.sign();
+      ledger.applyPartiesTransaction(
+        toParties(txn.transaction),
+        accountCreationFee_
+      );
+      return { wait: async () => {} };
+    },
+    async transaction(sender: SenderSpec, f: () => void) {
+      return createTransaction(sender, f);
+    },
+    applyJsonTransaction(json: string) {
+      return ledger.applyJsonTransaction(json, accountCreationFee_);
+    },
+    addAccount,
+    testAccounts,
+  };
+}
 
+function RemoteBlockchain(
+  graphqlEndpoint: string,
+  testAccounts = [] as {
+    publicKey: PublicKey;
+    privateKey: PrivateKey;
+  }[]
+): Testnet {
+  let accountCreationFee = UInt32.fromNumber(defaultAccountCreationFee);
+  Fetch.setGraphqlEndpoint(graphqlEndpoint);
   return {
-    accountCreationFee: () => UInt32.fromString(accountCreationFee_),
-    currentSlot,
-    getAccount,
-    transaction,
-    applyJsonTransaction,
-    addAccount,
+    accountCreationFee: () => accountCreationFee,
+    currentSlot() {
+      throw Error(
+        'currentSlot() is not implemented yet for remote blockchains.'
+      );
+    },
+    getAccount(publicKey: PublicKey) {
```

Comment:
or here is where we check the cache I guess?

Reply:
this only works when the cache is filled, to be synchronously callable